x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ilm/AbstractStepMasterTimeoutTestCase.java

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.elasticsearch.action.support.master.MasterNodeRequest;
+import org.elasticsearch.client.AdminClient;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.IndicesAdminClient;
+import org.elasticsearch.cluster.ClusterName;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.common.unit.TimeValue;
+import org.junit.Before;
+import org.mockito.Mockito;
+import org.mockito.stubbing.Stubber;
+
+import static org.elasticsearch.xpack.core.ilm.LifecycleSettings.LIFECYCLE_STEP_MASTER_TIMEOUT;
+import static org.hamcrest.Matchers.equalTo;
+
+public abstract class AbstractStepMasterTimeoutTestCase<T extends AsyncActionStep> extends AbstractStepTestCase<T> {
+
+    protected Client client;
+    protected AdminClient adminClient;
+    protected IndicesAdminClient indicesClient;
+
+    @Before
+    public void setup() {
+        client = Mockito.mock(Client.class);
+        adminClient = Mockito.mock(AdminClient.class);
+        Mockito.when(client.admin()).thenReturn(adminClient);
+        indicesClient = Mockito.mock(IndicesAdminClient.class);
+        Mockito.when(adminClient.indices()).thenReturn(indicesClient);
+    }
+
+    public void testMasterTimeout() {
+        checkMasterTimeout(TimeValue.timeValueSeconds(30),
+            ClusterState.builder(ClusterName.DEFAULT).metaData(MetaData.builder().build()).build());
+        checkMasterTimeout(TimeValue.timeValueSeconds(10),
+            ClusterState.builder(ClusterName.DEFAULT)
+                .metaData(MetaData.builder()
+                    .persistentSettings(Settings.builder().put(LIFECYCLE_STEP_MASTER_TIMEOUT, "10s").build())
+                    .build())
+                .build());
+    }
+
+    @SuppressWarnings("rawtypes")
+    private void checkMasterTimeout(TimeValue timeValue, ClusterState currentClusterState) {
+        IndexMetaData indexMetadata = getIndexMetaData();
+
+        Stubber checkTimeout = Mockito.doAnswer(invocation -> {
```

Comment:
I think we can do all this without mocking. I think by subclassing `NoOpClient` to check the executed request's master node timeout we can essentially do this check. I think this will be cleaner and less likely to break in the future than mocks.

Reply:
Yes, this is much better idea, changed